Upgrade to Hibernate ORM 7, Hibernate Reactive 3 and Hibernate Search 8

yrodiere opened the pull request:
This upgrades Quarkus to Hibernate ORM 7.

~~Based on #41359, which must be merged first.~~ => Done


- Closes: #45164
- Fixes: #43368
- Fixes: #38948
- Closes: #42247

Migration guide entry:

```adoc
== Jakarta Persistence / Hibernate ORM

=== Upgrade to Hibernate ORM 7.0 / Jakarta Persistence 3.2

The Quarkus extension for Hibernate ORM was upgraded to Hibernate ORM 7.0 / Jakarta Persistence 3.2.

Hibernate ORM 7.0 is for the most part backwards-compatible with Hibernate ORM 6.6, and comes with [many improvements and new features, as well as a switch from the LGPL license to ASL 2](https://docs.jboss.org/hibernate/orm/7.0/whats-new/whats-new.html).

However, as with any major version, a few breaking changes are to be expected. Below are the ones most likely to affect existing applications.

Refer to the https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html[Hibernate ORM 7.0 migration guide] for more information.

=== API changes

* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#defer-to-jpa[Various deprecated `Session` methods have been removed in favor of JPA equivalents]. In particular: `load(...)` => `getReference(...)`, `get(...)` => `find(...)`, `delete(...)` => `remove(...)`, `save(...)` => `persist(...)`, `update(...)` => `merge(...)`, `saveOrUpdate(...)` => `persist` (new entity)/`merge` (persisted, then detached entity).
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#removal-annotations[Various deprecated annotations methods have been removed in favor of JPA equivalents]. In particular: `@Table` = `@jakarta.persistence.Table`, `@Where`/`@WhereJoinTable` => `@SQLRestriction`/`@SQLJoinTableRestriction`, `@OrderBy` => `@SQLOrder` or `@jakarta.persistence.OrderBy`, `@Index` => `@jakarta.persistence.Index`, `@IndexColumn` => `@OrderColumn`.
* Hibernate's `@Cascade` and `CascadeType` are deprecated in favor of JPA equivalents (e.g. `@OneToMany(cascade = ...)`).
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#proxy-annotation[Proxy customization changed significantly]. If you were using `@Proxy, consider `@ConcreteProxy` instead, or taking advantage of `Session#getReference`/`Hibernate#unproxy` as necessary.
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#misc-api[`org.hibernate.Metamodel` was removed] in favor of `org.hibernate.metamodel.model.domain.JpaMetamodel`.
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#usertype[`UserType` and `CompositeType` had some method signature change] to no longer refer to internal/SPI types.
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#misc-spi[Various JDBC types were moved to an internal package]. Use `@JdbcTypeCode` instead of `@JdbcType` to map your attributes to these types.
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#lock-options[`LockOptions` was deprecated]; impacted methods have alternatives involving `LockMode`.

=== Behavior changes

* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#model-validation[Many misplaced mapping annotations will now result in an exception], instead of being ignored. This includes for example conflicting `@Basic`/`@ManyToOne` annotations on the same attribute, but also using attribute converters on `@Id`/`@Version`/associations/etc., which was never supported and was effectively ignored.
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#stateless-session-cache[`StatelessSession`] now takes advantage of the second-level cache.
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#stateless-session-jdbc-batching[`StatelessSession`] no longer complies with [quarkus.hibernate-orm.jdbc.statement-fetch-size](https://quarkus.io/guides/hibernate-orm#quarkus-hibernate-orm_quarkus-hibernate-orm-jdbc-statement-fetch-size). Instead, you can use explicit batch operations such as `insertMultiple()`.
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#create-query[Some ambiguous query strings are now disallowed]. This should mainly impact type-unsafe query creation, when not passing a class to `Session#createQuery`, or when using Panache to create a query. Relevant query strings are those without a `select` clause, but with a (non-fetch) join. For example `from Person p left join p.address` should become one of `from Person p`, `select p, a from p left join p.address a`, or `from p left join fetch p.address`, depending on the desired behavior.
* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#datetime-native[Native queries now return `java.time` types instead of `java.sql` types for date/time values] in the absence of specific type instructions.

=== DDL/schema changes

Quoting behavior, which suffered from a regression in Hibernate ORM 6.x, was fixed in 7.0 as part of https://hibernate.atlassian.net/browse/HHH-16516[HHH-16516]. Custom database initialization script, in particular on H2, may need to be adapted to avoid use of quotes.

The following changes will require updating the database schema, and manually updating data:

* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#array-mapping-changes-on-db2-sap-hana-sql-server-and-sybase-ase[Basic arrays (e.g. `String[]`) are now mapped to proper array structures in database] instead of being serialized as binary data. This behavior can be reverted by setting `quarkus.hibernate-orm.unsupported-properties."hibernate.type.preferred_array_jdbc_type"` to `VARBINARY`.

The following changes should not require a database schema or data update, as SQL generated by Hibernate should to work even with an older schema:

* https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#mysql-varchar[`char`/`Character` attributes are mapped to `varchar(1)`] instead of `char(1)`.
* On Oracle and Microsoft SQL Server, https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#ddl-implicit-datatype-timestamp[timestamp attributes have a different default sub-second precision].
* On Oracle, https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#float-mapping-changes-oracle[`float`/`double` attributes are mapped to `binary_float`/`binary_double`] instead of `float(p)`/`real`/`double precision` types.

=== Annotation processor artifact relocation 

Hibernate ORM 7.0 changes the Maven coordinates of its annotation processor responsible for the generation of the static metamodel and Jakarta Data repositories from `org.hibernate.orm:hibernate-jpamodelgen` to `org.hibernate.orm:hibernate-processor`. Make sure to update your `annotationProcessorPath` in Maven, or `annotationProcessor` instruction in Gradle.

Additionally, the annotation processor `org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor` is now deprecated
in favor of `org.hibernate.processor.HibernateProcessor`. If your build script was explicitly referring to the processor by name, e.g. with the `annotationProcessors` tag of `maven-compiler-plugin`, make sure to replace the deprecated processor name with the new one.

=== CDI restrictions for Hibernate ORM components

Some pluggable components, that are potentially retrieved before CDI is initialized, can no longer be CDI beans. This includes in particular:

* `org.hibernate.type.descriptor.jdbc.JdbcType`
* `org.hibernate.type.descriptor.java.BasicJavaType`
* `org.hibernate.type.descriptor.java.JavaType`
* `org.hibernate.usertype.UserType`
* `org.hibernate.usertype.UserCollectionType`
* `org.hibernate.usertype.CompositeUserType`
* `org.hibernate.type.descriptor.java.MutabilityPlan`
* `org.hibernate.metamodel.spi.EmbeddableInstantiator`
* `org.hibernate.generator.Generator` and its subclasses `org.hibernate.id.IdentifierGenerator`, `org.hibernate.generator.AnnotationBasedGenerator`, etc.
* `org.hibernate.envers.RevisionListener`

If you need to provide a custom implementation of such components, and need access to CDI within that implementation, consider retrieving other beans when first needed, using `Arc.container().instance(OtherBean.class).get()`.
 
== Elasticsearch

Default connection pool sizes have been bumped https://hibernate.atlassian.net/issues/HSEARCH-5045[to provide a better experience with Hibernate Search], in particular when mass indexing:

* `quarkus.elasticsearch.max-connections` now defaults to `40` instead of `20`.
* `quarkus.elasticsearch.max-connections-per-route` now defaults to `20` instead of `10`.

=== Upgrade to Hibernate Search 8.0

The Quarkus extensions for Hibernate Search were upgraded to Hibernate Search 8.0.

Hibernate ORM 8.0 is for the most part backwards-compatible with Hibernate Search 7.2, and comes with https://hibernate.org/search/releases/8.0/#whats-new[many improvements and new features], in particular a https://hibernate.org/search/releases/8.0/#static-metamodel[static metamodel]

However, as with any major version, a few breaking changes are to be expected.
Below are the ones most likely to affect existing applications.

Refer to the the https://docs.jboss.org/hibernate/search/8.0/migration/html_single/[Hibernate Search 8.0 migration guide] for more information.

* Deprecated classes have been removed in favor of their recommended alternatives, in particular `org.hibernate.search.mapper.orm.massindexing.MassIndexingMonitor` => `org.hibernate.search.mapper.pojo.massindexing.MassIndexingMonitor`.
* `MassIndexingMonitor#addToTotalCount` is now deprecated in favor of the more flexible `MassIndexingMonitor#typeGroupMonitor(..)`
* `multi()` methods in the projection DSL are deprecated in favor of `.list()`/`.set()`/`sortedSet()`/`collector(...)`.
* Logging categories have been overhauled. See https://docs.jboss.org/hibernate/search/8.0/reference/en-US/html_single/#logging-categories-aggregated[here for a list of available categories]. In particular, the `org.hibernate.search.elasticsearch.request` logging category, used for logging every request sent to Elasticsearch, has been renamed to `org.hibernate.search.elasticsearch.client.request`.
* Many Search DSL interfaces now have an extra `SR` type argument to accomodate for the new static metamodel. This mostly affects `*Step` interfaces which should generally not be used in application code.
* The format of mass indexing logs has changed -- for the better, as more information is presented in a more condensed format.

== Dev Services

Several Dev Services default images have been updated:

* Elasticsearch from 8.18 to 9.0
* OpenSearch from 2.16 to 3.0

NOTE: You can configure Quarkus explicitly to use a specific image for each Dev Service, e.g. see https://quarkus.io/guides/elasticsearch-dev-services#configuring-the-image[here for Elasticsearch/OpenSearch].
```

Status / next steps:

- [x] Make it compile
- [x] Make Java-only, ORM-only tests pass
- [x] Make Kotlin ORM-only tests pass (requires https://github.com/hibernate/hibernate-models/issues/99)
- [x] Upgrade to compatible Search/Reactive versions
- [x] Make Search/Reactive tests pass
- [x] If necessary, upgrade to compatible Validator version
- [x] Inspect commits for temporary workarounds (mentions of `HHH-`) and remove them if they are no longer necessary -- see in particular https://hibernate.atlassian.net/browse/HHH-19033
- [x] Inspect commits + [upstream migration guide](https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html) and write migration notes for Quarkus (ping Guillaume so he can assess the situation)
- [ ] Inspect commits + [upstream migration guide](https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html) and write [OpenRewrite recipes in `quarkus-update`](https://github.com/quarkusio/quarkus-updates/tree/main/recipes/src/main/resources/quarkus-updates/core)
- [x] Upgrade to Final versions
- [x] Link to fixed Quarkus issues in this PR -- see https://github.com/quarkusio/quarkus/issues?q=is%3Aissue%20is%3Aopen%20label%3Aarea%2Fhibernate-orm%2Carea%2Fhibernate-reactive%2Carea%2Fhibernate-search%20label%3Akind%2Fbug-thirdparty
- [x] Check whether #45263 gets fixed by the Hibernate Reactive update, because Hibernate Reactive is now much more similar to ORM when it comes to DB version setting/checking. If relevant, consider documenting DB version setting/checking for Hibernate Reactive. => Moved to #43764
- [x] Make sure to support injecting `jakarta.persistence.SchemaManager` -- see https://github.com/quarkusio/quarkus/pull/47792
- [x] Take Elasticsearch upgrades into account in migration guide

yrodiere committed:
Upgrade to Hibernate ORM 7.0.0.Alpha3

yrodiere committed:
Upgrade to Hibernate ORM 7.0.0.Beta1

yrodiere committed:
Disallow CDI beans impacting metadata creation

Because:

1. We currently disallow this for identifier generators only
   through a custom initializer for the identifier generator
   factory service, but that service is disappearing in Hibernate
   ORM 7.0.0.Beta1, leaving us only the setting `hibernate.cdi.extensions`.
   to disallow CDI for identifier generators -- and anything else
   that impacts metadata creation.
2. This is needed for https://github.com/quarkusio/quarkus/issues/40897,
   which will move more of metadata creation to build time -- where CDI
   is just not available.
3. Implementations of affected components needing access to CDI at
   runtime (so not during metadata creation) can still do so by
   calling `Arc.container()` to retrieve the relevant beans.

gsmet commented:
Should we try to coordinate the Jakarta EE 11 updates?

yrodiere commented:
> Should we try to coordinate the Jakarta EE 11 updates?

Quarkus definitely should plan for that, yes. I don't know the full implications for Quarkus, but I think at least CDI, Narayana and Validator (so probably EL too?) will need to be upgraded before or at the same time as we upgrade to Hibernate ORM 7.0.
Upgrading every Jakarta component to Jakarta 11 would probably be simpler than trying to mix Jakarta 10 and 11.

Fair warning though:

1. ORM 7.0 is not ready yet -- there's no clear date set for the release at the moment.
2. I'm only just getting back to this, trying to get things to compile (:heavy_check_mark:)  and basic tests to pass (:x:), so that I can report any bugs upstream.

I'll set a reminder to ping you when I have a clearer view of what works and what doesn't in Quarkus itself.

yrodiere committed:
Make Hibernate ORM and jdbc-derby extensions work without any other required dependencies

With Derby moving to hibernate-community-dialects, Quarkus will need to
add this dependency automatically.

yrodiere commented:
I identified the Kotlin failures as a bug in hibernate-models: https://github.com/hibernate/hibernate-models/issues/99
Looking into the other failures...

yrodiere commented:
The XML mapping issue seems caused by https://hibernate.atlassian.net/browse/HHH-18662.

yrodiere commented:
Alright, I think I did all I could at this point. I will need new Beta releases of Hibernate ORM / Search / Reactive to go further.

yrodiere committed:
Upgrade to Hibernate ORM 7.0.0.Beta3 / Reactive 3.0.0.Beta1

yrodiere commented:
I pushed an upgrade to ORM 7.0.0.Beta2, but now the build is failing even earlier.

Status:

1. ORM needs a Beta3, because Beta2 has incorrect dependencies or a missing artifact:
    ```
    [ERROR] Failed to execute goal on project quarkus-hibernate-orm-derby: Could not resolve dependencies for project io.quarkus:quarkus-hibernate-orm-derby:jar:999-SNAPSHOT
    [ERROR] dependency: org.hibernate.orm:hibernate-scan-jandex:jar:7.0.0.Beta2 (runtime)
    [ERROR]         Could not find artifact org.hibernate.orm:hibernate-scan-jandex:jar:7.0.0.Beta2 in central (https://repo.maven.apache.org/maven2)
    ```
2. Search needs an 8.0.0.Alpha1
3. Reactive needs an Alpha/Beta as well... which may require further changes to ORM.

yrodiere commented:
I upgraded my branch to Hibernate ORM 7.0.0.Beta3 / Hibernate Reactive 3.0.0.Beta1, let's see what CI thinks... :crossed_fingers: 

@FroMage FYI ^

yrodiere committed:
Adapt QuarkusStrategySelectorBuilder to changes in StrategySelectorBuilder

Some of these changes probably pre-date Hibernate ORM 7.0,
we just forgot to implement them.

yrodiere committed:
Work around https://hibernate.atlassian.net/browse/HHH-19033

yrodiere commented:
Current status: native compilation is broken, but problems have been identified:

1. One is fixed and just needs a new release of ORM: https://github.com/hibernate/hibernate-orm/pull/9589
2. The other needs a fix _and_ a new release of ORM: https://hibernate.atlassian.net/browse/HHH-18974

yrodiere committed:
Upgrade to Hibernate ORM 7.0.0.Beta4 / Reactive 3.0.0.Beta2

Co-Authored-By: Jan Schatteman <jrenaat@users.noreply.github.com>

@ Conflicts:
@	pom.xml

yrodiere committed:
Upgrade to Hibernate Search 8.0.0.Alpha1

And remove Hibernate Commons Annotations, which isn't needed by any
Hibernate project anymore.

yrodiere committed:
Upgrade to Hibernate Search 8.0.0.Alpha2

yrodiere commented:
Time for a status update!

* Currently using ORM 7.0.0.Beta4 , Search 8.0.0.Alpha2, Reactive 3.0.0.Beta2
* Had some problems with Panache related to [HHH-18584](https://hibernate.atlassian.net/browse/HHH-18584), but they're solved in blocking Panache, and I'm hoping my last push will solve them in reactive Panache as well.
* Hibernate Reactive native compilation fails [because of logging](https://hibernate.zulipchat.com/#narrow/channel/205413-hibernate-reactive-dev/topic/Reactive.203.20upgrade.20in.20Quarkus), but that should be solved in the next release
* Hibernate Search native compilation fails [because of explicit references to Jandex classes](https://hibernate.zulipchat.com/#narrow/channel/132092-hibernate-search-dev/topic/Hibernate.20Search.208.20in.20Quarkus/near/506677344), but that should be solved in the next release
* Envers is giving me trouble because of [HHH-19258](https://hibernate.atlassian.net/browse/HHH-19258), but I think I'll just implement a nasty workaround for now.

yrodiere committed:
Take into account feature removal HHH-18584 in tests

See https://docs.jboss.org/hibernate/orm/7.0/migration-guide/migration-guide.html#create-query
See https://hibernate.atlassian.net/browse/HHH-18584

yrodiere commented:
> Had some problems with Panache related to [HHH-18584](https://hibernate.atlassian.net/browse/HHH-18584), but they're solved in blocking Panache, and I'm hoping my last push will solve them in reactive Panache as well.

False hopes, that won't work. I gave up on fixing Panache, and stuck to fixing the queries we generate for Spring Data (add a `select` clause). Any query in either Spring Data or Panache that omits the `select` clause will still potentially hit the consequences of HHH-18584.

yrodiere commented:
Current status:

* Currently using ORM 7.0.0.Beta4 , Search 8.0.0.Alpha2, Reactive 3.0.0.Beta2
* Panache problems related to [HHH-18584](https://hibernate.atlassian.net/browse/HHH-18584) are worked around in the specific case of named-based methods in the Spring Data implementation (based on Panache). Direct uses will still suffer from a breaking change which we'll need to document. @gsmet I checked, and none of the examples in the Panache docs. seem to be affected.
* Hibernate Reactive native compilation fails [because of logging](https://hibernate.zulipchat.com/#narrow/channel/205413-hibernate-reactive-dev/topic/Reactive.203.20upgrade.20in.20Quarkus), but that should be solved in the next release
* Hibernate Search native compilation fails [because of explicit references to Jandex classes](https://hibernate.zulipchat.com/#narrow/channel/132092-hibernate-search-dev/topic/Hibernate.20Search.208.20in.20Quarkus/near/506677344), but that should be solved in the next release
* Apart from that, the whole build passes, native tests included.

yrodiere committed:
Upgrade to Hibernate ORM 7.0.0.Beta5 / Hibernate Reactive 3.0.0.Beta3

yrodiere committed:
Upgrade to Hibernate Search 8.0.0.Alpha3

yrodiere committed:
Upgrade to Hibernate ORM 7.0.0.CR1 / Hibernate Reactive 3.0.0.CR1 / Hibernate Search 8.0.0.Beta1

yrodiere committed:
Upgrade to Hibernate ORM 7.0.0.CR2 / Hibernate Reactive 3.0.0.CR2

yrodiere committed:
Remove some dead code in Panache using Hibernate ORM internals

These internals are changing in Hibernate ORM 7

yrodiere committed:
Upgrade to Hibernate ORM 7.0.0.Final / Hibernate Reactive 3.0.0.Final / Hibernate Search 8.0.0.CR1

gsmet commented on pom.xml:
```diff
@@ -71,14 +71,14 @@
         <bytebuddy.version>1.15.11</bytebuddy.version> <!-- version controlled by Hibernate ORM's needs -->
-        <hibernate-commons-annotations.version>7.0.3.Final</hibernate-commons-annotations.version> <!-- version controlled by Hibernate ORM's needs -->
-        <hibernate-reactive.version>2.4.8.Final</hibernate-reactive.version> <!-- highly sensitive to Hibernate ORM upgrades -->
+        <hibernate-models.version>1.0.0.CR3</hibernate-models.version> <!-- version controlled by Hibernate ORM's needs -->
```
Is the final version of ORM 7 using a CR3 of hibernate-models?

I don't see a Final on Central so I suppose it's the case but it's a bit odd?

yrodiere replied:
Yeah it's odd -- I'll bring this up with Steve.

gsmet commented on extensions/hibernate-search-standalone-elasticsearch/runtime/pom.xml:
```diff
@@ -28,6 +28,8 @@
         <dependency>
             <groupId>org.hibernate.search</groupId>
             <artifactId>hibernate-search-mapper-pojo-standalone</artifactId>
+            <exclusions>
+            </exclusions>
```
Probably should be dropped?

gsmet commented on extensions/spring-data-jpa/deployment/src/test/resources/import_customers.sql:
```diff
@@ -1,11 +1,11 @@
-INSERT INTO "Customer$Country" (id, name, isocode)
+INSERT INTO Customer$Country (id, name, isocode)
```
Could you explain why this change?

Asking because it could impact users.

yrodiere replied:
It's discussed in (too) great details here: https://hibernate.zulipchat.com/#narrow/channel/132094-hibernate-orm-dev/topic/Uppercase.20table.20names.20with.20H2.20in.20ORM.207.3F/with/506173896

Bottom line: it's a behavior change to fix a regression. 5.x behaved a certain way, 6.x another, and in 7.x we're back to the 5.x behavior.

I agree it'd worth mentioning in the migration guide.

gsmet reviewed (commented):
This looks good overall and it's nice to see it close to being ready.

I added a few comments/questions.

marko-bekhta committed:
Upgrade to Elasticsearch client 9.0.1

to match the one used in Hibernate Search

marko-bekhta committed:
Bump Elasticsearch/OpenSearch containers to latest-compatible with Hibernate Search

yrodiere committed:
Restore proxy-related tests using `@ConcreteProxy` instead of `@Proxy(proxyClass = ...)`

This is the recommended approach to migrate.
That, or using `Hibernate.unproxy()` if the impact of
`@ConcreteProxy` on SQL performance is not acceptable.

See https://hibernate.zulipchat.com/#narrow/channel/132094-hibernate-orm-dev/topic/HHH-18194/near/467597335

yrodiere committed:
Expose `jakarta.persistence.SchemaManager` as a CDI bean

yrodiere committed:
Upgrade to Hibernate Models 1.0.0

This means we' temporarily misaligned with the version used in ORM
7.0.0.Final, but that version is a CR3 so we don't want it in Quarkus,
and anyway to misalignment should be fine as no change was necessary in
ORM for the upgrade:
https://github.com/hibernate/hibernate-orm/pull/10282/files

yrodiere committed:
Upgrade to Hibernate Search 8.0.0.Final / Elasticsearch client 9.0.2

yrodiere committed:
Bump the maximum number of connections for Elasticsearch

To align on Hibernate Search: https://docs.jboss.org/hibernate/search/8.0/migration/html_single/#configuration

See here for the rationale: https://hibernate.atlassian.net/issues/HSEARCH-5045

yrodiere committed:
Remove unnecessary log disabling for Hibernate Search

Hibernate Search complies with jboss.log-version, see io.quarkus.runtime.logging.JBossVersion

marko-bekhta committed:
Use hibernate-processor instead of the hibernate-jpamodelgen

gsmet commented on build-parent/pom.xml:
```diff
@@ -73,12 +73,12 @@
 
         <!-- Defaults for integration tests -->
-        <elasticsearch-server.version>8.18.2</elasticsearch-server.version>
+        <elasticsearch-server.version>9.0.1</elasticsearch-server.version>
```
I see `9.0.1` here and `9.0.2` above?

yrodiere replied:
This one is the Elasticsearch server (container) version, the other one is the client. They don't need to be aligned perfectly.

But it's better to be on the latest micro, so I'll align.

gsmet commented on docs/src/main/asciidoc/hibernate-orm.adoc:
```diff
@@ -1704,11 +1707,14 @@ you must configure the annotation processor in your build tool:
             <!-- other processors that may be required by your app -->
         </annotationProcessorPaths>
+        <annotationProcessors>
+            <annotationProcessor>org.hibernate.processor.HibernateProcessor</annotationProcessor>
```
Why do we need this now?

marko-bekhta replied:
oh that's on me 🙈, previous processor `org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor` is deprecated in favour of this "new" one `org.hibernate.processor.HibernateProcessor` (but essentially it is just a package/class name change as `JPAMetaModelEntityProcessor extends HibernateProcessor`) and I missed that only the new one is referenced in the `meta-inf/services` so while being explicit about which processor to apply does not hurt, it is also redundant here 🙈

yrodiere replied:
Removed

gsmet commented on docs/src/main/asciidoc/hibernate-orm.adoc:
```diff
@@ -1704,11 +1707,14 @@ you must configure the annotation processor in your build tool:
             <path>
                 <groupId>org.hibernate.orm</groupId>
-                <artifactId>hibernate-jpamodelgen</artifactId>
+                <artifactId>hibernate-processor</artifactId>
```
We will probably need to write an OpenRewrite recipe for this one.

marko-bekhta replied:
Do we already have the recipes for this version? I've only seen some for the previous ones, when making the change but maybe I just missed it 🙈

gsmet reviewed (approved):
I added one comment and some questions but I would rather merge early than late so feel free to ignore and adjust later :).

marko-bekhta commented:
```asciidoc
=== Annotation processor artifact relocation 

Hibernate ORM 7.0 change the coordinates of its annotation processor responsible for the generation of the static metamodel and Jakarta Data repositories 
from `org.hibernate.orm:hibernate-jpamodelgen` to `org.hibernate.orm:hibernate-processor`.
Additionally, the annotation processor `org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor` is now deprecated
in favour of `org.hibernate.processor.HibernateProcessor`. Make sure to replace the deprecated processor
if it was explicitly mentioned in the build script, e.g. in the `annotationProcessors` of the Maven's compiler plugin.
```

to add to the guide

yrodiere commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/util/HibernateProcessorUtil.java:
```diff
@@ -174,7 +173,6 @@ public static void configureProperties(QuarkusPersistenceUnitDescriptor desc, Hi
         int batchSize = firstPresent(config.fetch().batchSize(), config.batchFetchSize()).orElse(defaultBatchSize(reactive));
         if (batchSize > 0) {
             desc.getProperties().setProperty(AvailableSettings.DEFAULT_BATCH_FETCH_SIZE, Integer.toString(batchSize));
-            desc.getProperties().setProperty(AvailableSettings.BATCH_FETCH_STYLE, BatchFetchStyle.PADDED.toString());
```
The setting has been deprecated since 6.0, and was removed in 7.0. AFAICT Hibernate ORM 6 sometimes uses padding, and sometimes dynamic sizing. See here if you want to dig: https://github.com/hibernate/hibernate-orm/blob/9c5cf75111885c2535d1373671817e89f4e896ab/hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java#L4459-L4471

yrodiere commented:
Thank you! I added the relevant information to the migrationg guide: https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.24#jakarta-persistence-hibernate-orm

michalvavrik commented:
@yrodiere the note you added https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.24#jakarta-persistence-hibernate-orm has some typo that causes wrong link rendering in:

```
and comes with [many improvements and new features, as well as a switch from the LGPL license to ASL 2](https://docs.jboss.org/hibernate/orm/7.0/whats-new/whats-new.html).
```

please fix the link, thanks
